packages/mcp/src/secrets/secret-manager.ts

```diff
@@ -0,0 +1,234 @@
+   *
+   * @param providers - Array of secret providers to use for resolution
+   * @param registry - Optional registry for additional provider management
+   * @param options - Configuration options for the manager
+   */
+  constructor(
+    providers: ISecretProvider[] = [],
+    registry?: ISecretProviderRegistry,
+    options: SecretManagerOptions = {},
+  ) {
+    this.providers = [...providers]; // Create a defensive copy
+    this.registry = registry;
+    this.cacheTtl = options.cacheTtl ?? 300000; // 5 minutes default
+    this.logger = options.logger ?? defaultLogger;
+  }
+
+  /**
+   * Resolves secrets from all registered providers.
+   *
+   * Calls resolveSecrets() on all providers and merges the results.
+   * Later providers in the array override values from earlier providers.
+   * Provider errors are logged but do not stop the resolution process.
+   *
+   * @returns A promise that resolves to the merged secrets from all providers
+   */
+  async resolveSecrets(): Promise<Record<string, string>> {
+    // Check cache first
+    if (this.cache && this.isCacheValid()) {
+      return { ...this.cache.secrets }; // Return defensive copy
+    }
+
+    const allSecrets: Record<string, string> = {};
+    const allProviders = this.getAllProviders();
+
+    // Resolve secrets from each provider
+    for (const provider of allProviders) {
+      try {
+        const providerSecrets = await provider.resolveSecrets();
+
+        // Merge with later providers overriding earlier ones
+        Object.assign(allSecrets, providerSecrets);
+      } catch (error) {
+        // Log error but continue with other providers
+        const providerName = provider.getName();
+        this.logger.error(
+          `Failed to resolve secrets from provider '${providerName}'`,
+          error,
+          { providerName },
+        );
+      }
+    }
+
+    // Update cache if caching is enabled
+    if (this.cacheTtl > 0) {
+      this.cache = {
+        secrets: { ...allSecrets }, // Store defensive copy
+        resolvedAt: new Date(),
+        ttl: this.cacheTtl,
+      };
+    }
+
+    return allSecrets;
+  }
+
+  /**
+   * Adds a new provider to the manager.
+   *
+   * The provider will be added to the end of the provider list,
+   * giving it the highest precedence for secret resolution.
+   *
+   * @param provider - The secret provider to add
+   */
+  addProvider(provider: ISecretProvider): void {
+    this.providers.push(provider);
+    this.invalidateCache();
+  }
+
+  /**
+   * Removes the first provider from the manager by name.
+   *
+   * @param name - The name of the provider to remove (from getName())
+   * @returns true if a provider was removed, false if no provider with that name was found
+   */
+  removeProvider(name: string): boolean {
+    const index = this.providers.findIndex(
+      (provider) => provider.getName() === name,
+    );
+
+    if (index === -1) {
+      return false;
+    }
+
+    this.providers.splice(index, 1);
```

Comment:
`removeProvider` bails after deleting the first match, but the manager still allows duplicate provider registrations (integration tests register multiple inline providers with the same `name`). A caller trying to unregister "inline" removes only the first entry while later duplicates stay active, so secrets they expected to drop continue to leak through. Please either filter out **all** providers with the requested name or enforce unique provider names when adding providers so the API behaves predictably.

Reply:
/fix